### ISSUE TITLE
Fix S4158 FP: HashSet.UnionWith should not be seen as iteration or access

### ISSUE THREAD

**Issue by omarjuul** [SonarSource/sonar-dotnet#7764]
### Description

`HashSet.UnionWith` causes S4158 on an empty `HashSet`. It should not.

### Repro steps

```
var foo1 = new[] { 1, 2 };
var foo2 = new[] { 2, 3 };

bool condition1 = true;
bool condition2 = true;

HashSet<int> set = new();
if (condition1)
{
    set.UnionWith(foo1);
}
if (condition2)
{
    set.UnionWith(foo2);
}
```

### Expected behavior

S4158 is not issued as the empty `HashSet` is part of the logic and `UnionWith` is used to add to this `HashSet`. `AddRange` or a similar method is not available on `HashSet<T>`.

### Actual behavior

S4158 is issued.

### Known workarounds

Suppress the issued warning.

### Related information
`SonarAnalyzer` version `9.7.0.75501`

**Comment by cristian-ambrosini-sonarsource** [comment]
Hi @omarjuul, I confirm this as a FP, I added a reproducer in our repository. Thanks for your contribution!

**Comment by cristian-ambrosini-sonarsource** [comment]
Hi @omarjuul!
After discussing this false positive with my colleagues, I have reconsidered my initial response. Upon further evaluation, we believe that the rule is functioning correctly. Although `UnionWith` may require some code refactoring, it is still appropriate to notify the user that there is a more optimal approach available.
For example, the code you reported can be rewritten as such:

```cs
var foo1 = new[] { 1, 2 };
var foo2 = new[] { 2, 3 };

bool condition1 = true;
bool condition2 = true;

HashSet<int> set;
if (condition1)
{
    set = foo1; // Use assignment, instead of UnionWith
}
if (condition2)
{
    set = foo2; // Use assignment, instead of UnionWith
}
```

I'm happy to answer any questions you might have or provide further clarification.

**Comment by omarjuul** [comment]
Hi @cristian-ambrosini-sonarsource 
thanks for your response. I think you simplified the example a bit too much.
My initial code example leads to ```set == { 1, 2, 3 }```.
Your modified example results in ```set == { 2, 3 }```.

This is with the supplied values of `true` for both conditions. `condition1` and `condition2` can be `true` and/or `false` independently from each other.

To get the original behavior (or at least end result), I think something like the following would be needed:
```C#
var foo1 = new[] { 1, 2 };
var foo2 = new[] { 2, 3 };

bool condition1 = true;
bool condition2 = true;

HashSet<int> set;
if (condition1 && condition2)
{
    set = foo1;
    set.UnionWith(foo2); // maybe this should be done a bit differently, depending on whether `foo1` and/or `foo2` are used further on
}
else if (condition1)
{
    set = foo1; // Use assignment, instead of UnionWith
}
else if (condition2)
{
    set = foo2; // Use assignment, instead of UnionWith
}
else
{
    set = new();
}
```
This is overly verbose and not a coding style I like. I think I would work around this by suppressing S4158 and using the initial code. Which is a shame because it's a useful rule otherwise.

Maybe I'm missing some other option of achieving the result I want. Either way I still think this actually *is* a false positive: `UnionWith` can semantically be used in the same way as `AddRange` and should be available on a (known to be) empty collection.

**Comment by cristian-ambrosini-sonarsource** [comment]
Hi @omarjuul, 
I apologize for the confusion caused by my previous response, thank you for pointing out the mistake.
Here's an updated code snippet that should address your requirements:

```cs
    var foo1 = new[] { 1, 2 };
    var foo2 = new[] { 2, 3 };

    bool condition1 = true;
    bool condition2 = true;

    HashSet<int> set = new();
    if (condition1)
    {
        set = new(foo1);
    }
    if (condition2)
    {
        set.UnionWith(foo2); // S4158 does not raise here
    }
```
I understand that using 2 `UnionWith` calls might result in cleaner code. However, please let me know if this solution meets your requirements or if you have any other suggestions.

**Comment by AtomicTroop** [comment]
Hi, I believe this issue should be reopened, as the suggested workaround isn't equivalent from a memory management standpoint.

Particularily, consider the following (somewhat specific and convoluted) usecase:


In this usecase, we are concerned with minimizing memory allocations. This is a platform restriction/requirement.

UpdateValues is run periodically to poll the state of an external library for key-value mappings. The external library manages its own state internally.

The values in this mapping need to go through an expensive transformation process before storage in this class instance's local dictionary. Thus, we don't want to apply the transformation again on values that already exist in the local dictionary.

Additionally, any mappings that have disappeared from the external library's state need to be removed from the local dictionary's mappings. To achieve this without allocations, an additional HashSet is created once and reused to track which Guid keys are now stale and should be removed.
```csharp

private Dictionary<Guid, FooValue>   _fooValues = new();
private HashSet<Guid>                _oldGuids      = new();

private void UpdateValues() {
    _oldGuids.Clear();
    _oldGuids.UnionWith(_fooValues.Keys);

    foreach (var (guid, value) in SomeExternalLibrary.ValueMap) {
        if (_oldGuids.Contains(guid))
            _oldGuids.Remove(guid);
        else
            _fooValues.Add(guid, SomeExpensiveTransformation(value));
    }

    foreach (var guid in _oldGuids) {
        _fooValues.Remove(guid);
    }
}
```

This would seem to be a perfectly valid usecase where S4158 is triggered while the offered workarounds compromise the domain specific requirements.

**Comment by cristian-ambrosini-sonarsource** [comment]
Hi @AtomicTroop!
I'm not sure I get the concept so I'll try to summarize, please correct me if I'm wrong. From what I understood in your snippet S4158 is raising
```cs
_oldGuids.Clear();
_oldGuids.UnionWith(_fooValues.Keys); // <-- HERE
```
and you are saying that doing:
```cs
_oldGuids = new(_fooValues.Keys);
```
is worse, from a memory perspective. Is this correct?

**Comment by AtomicTroop** [comment]
That's the gist of it, yes.
Wouldn't creating a new instance rather than clearing the previous one allocate new memory and leave the previous instance to be garbage collected?

I do understand that this is not a concern in most applications.

**Comment by cristian-ambrosini-sonarsource** [comment]
Yes, in this particular case using `.Clear()` removes the elements from the HashSet but it doesn't release the memory that was allocated for those elements. This can be beneficial only if you're going to add a similar amount of data back into the HashSet, as it won't need to resize.
On the other hand, `_oldGuids = new(_fooValues.Keys)` creates a new object and the memory for the old hashSet will be released (once GC runs).
I think both operations might have different impacts on memory depending on the context.

IMO raising S4158 on "UnionWith" has still more advantages than disadvantages, and improving the rule to grasp the context is not feasible. In those specific instances, it's simply okay to "won't fix" the FP, for the reason you explained.

**Comment by Tim-Pohlmann** [comment]
I collected multiple use cases for when `UnionWith` on an empty collection is fine:
- Instantiating a new `HashSet` would be required but is not desirable for performance reasons.
- The set is a `readonly` field (or something else that cannot be assigned). #8483
- The set is a non-`ref` method parameter:
```csharp
void Method(ISet set)
{
    set.Clear();
    set.UnionWith(GetItems());
}
```

There are likely a few more cases where it is the best or only approach. Therefore, I have reconsidered this issue and decided that **not** raising on `UnionWith` is desirable.
Coincidentally, this was implemented already in #8712.
